### ISSUE TITLE
`__attribute__((pure))`/`__attribute__((const))` is a massive unchecked footgun

### ISSUE THREAD

**Issue by LebedevRI** [llvm/llvm-project#58798]
This originates from https://oss-fuzz.com/testcase-detail/5834828872548352

Consider (somewhat reduced from the original source):
```
#include <vector>

std::vector<int> handle() {
  std::vector<int> v(42, 42); // this somehow leaks
  return v;
}

__attribute__((pure)) // double yikes
std::vector<int> footgun(int argc) {
  std::vector<int> v(24, 24);
  if(argc != 42)
    throw int(0); // yikes
  return v;
}

int main(int argc, char* argv[]) {
    try {
        auto v = handle();
        auto v2 = footgun(argc);
    } catch(...) {}
    return 0;
}
```
https://godbolt.org/z/zdavdKnfa

Not a single diagnostic is triggered by this.
Yet this is completely broken.

In oss-fuzz issue, this manifested as an obscure leak,
exception got optimized away, so this took a bit to understand...

```
    // 'const', 'pure' and 'noalias' attributed functions are also nounwind.
    if (TargetDecl->hasAttr<ConstAttr>()) {
      FuncAttrs.addAttribute(llvm::Attribute::ReadNone);
      FuncAttrs.addAttribute(llvm::Attribute::NoUnwind);
      // gcc specifies that 'const' functions have greater restrictions than
      // 'pure' functions, so they also cannot have infinite loops.
      FuncAttrs.addAttribute(llvm::Attribute::WillReturn);
    } else if (TargetDecl->hasAttr<PureAttr>()) {
      FuncAttrs.addAttribute(llvm::Attribute::ReadOnly);
      FuncAttrs.addAttribute(llvm::Attribute::NoUnwind);
      // gcc specifies that 'pure' functions cannot have infinite loops.
      FuncAttrs.addAttribute(llvm::Attribute::WillReturn);
```

Which means functions with these attributes *can* can call non-returning functions,
or functions that throw exceptions, but the exception must be handled within the function.
Yet we fail to catch that at all, even in UBSan.

I think at least 3 things are missing:
* we should infer `nothrow` attribute (and manifest in AST!) upon seeing `const`/`pure`.
  that would catch the simplified snippet: https://godbolt.org/z/nM6PM6Msc
* we lack `willthrow` clang attribute. (maybe also `maythrow`?)
  * we should add it, infer it from `throw` stmts
  * and propagate it in obvious cases through call stack
* we should diagnose calls to `willthrow` functions from `nothrow` functions :)

CC @AaronBallman @regehr

**Comment by LebedevRI** [comment]
Posted UBSan part: https://reviews.llvm.org/D137381

**Comment by nikic** [comment]
FWIW GCC does not consider pure to imply nounwind, there's a couple open bugs about that already. Might be worth changing the Clang semantics to match.

**Comment by AaronBallman** [comment]
First off, I totally agree these are footguns and I really appreciate any efforts on diagnosing them!

Since we're talking a little bit about changing semantics, one thing we should also keep in mind is that these got kinda, sorta standardized but not in the same way as the GNU-style attributes: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2956.htm

From the paper:
> The following two attributes then combine these to approximate the GCC attributes pure and const
> * reproducible: effectless and idempotent – close to GCC pure
> * unsequenced: stateless, effectless, idempotent, and independent – close to GCC const

So hopefully, if we do decide to change semantics, we can avoid making support for those harder.

**Comment by FrankHB** [comment]
> FWIW GCC does not consider pure to imply nounwind, there's a couple open bugs about that already. Might be worth changing the Clang semantics to match.

I've reported one years ago: #42620.

Any others?
